tests/Main.hs

```diff
@@ -28,6 +31,26 @@ test6 _ _ = natVal (Proxy :: Proxy ((n^m)+(n*m)))
 test11 :: forall m . (KnownNat m) => Proxy m -> Integer
 test11 _ = natVal (Proxy @ (m*m))
 
+test12 :: forall m . (KnownNat (m+1)) => Proxy m -> Integer
+test12 = natVal
+
+test13 :: forall m . (KnownNat (m+3)) => Proxy m -> Integer
+test13 = natVal
+
+test14 :: forall m . (KnownNat (4+m)) => Proxy (7+m) -> Integer
+test14 = natVal
+
+type family Foo (m :: Nat) = (result :: Nat) | result -> m
```

Comment:
perhaps if we write:

```
type family Foo (m :: Nat) = (result  :: Nat) | result -> m where
  Foo 0 = 0
  Foo m = m
```

then maybe the later unsafeCoerce's are not needed?

p.s. I don't know if closed type families allow type family dependency annotations